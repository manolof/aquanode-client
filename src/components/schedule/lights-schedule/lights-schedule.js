import PropTypes from 'prop-types';
import React, { Component } from 'react';

class LightsSchedule extends Component {

	type = 'lights';

	constructor(props) {
		super(props);
		this.state = {
			red: '',
			green: '',
			blue: '',
			white: '',
		};
	}

	handleChange = (e) => {
		e.persist();
		const value = e.target.value;
		this.setState(state => ({
			...state,
			[e.target.name]: value,
		}));
	};

	render() {
		const { schedule, resetSchedule, setSchedule } = this.props;
		const { red, green, blue, white } = this.state;

		return (
			<section className="schedule schedule__lights">
				<h4>Lights</h4>

				<article className="schedule__values">
					{schedule}
				</article>

				<form className="schedule__form">
					<label>
						R
						<input
							type="number"
							name="red"
							min="0"
							max="255"
							value={red}
							onChange={this.handleChange}
						/>
					</label>
					<label>
						G
						<input
							type="number"
							name="green"
							min="0"
							max="255"
							value={green}
							onChange={this.handleChange}
						/>
					</label>
					<label>
						B
						<input
							type="number"
							name="blue"
							min="0"
							max="255"
							value={blue}
							onChange={this.handleChange}
						/>
					</label>
					<label>
						W
						<input
							type="number"
							name="white"
							min="0"
							max="255"
							value={white}
							onChange={this.handleChange}
						/>
					</label>
				</form>

				<footer className="schedule__footer">
					<button className="schedule__reset" onClick={() => resetSchedule(this.type)}>RESET</button>
					<button className="schedule__set" onClick={() => setSchedule(this.type, this.state)}>SET</button>
				</footer>
			</section>
		);
	}
}

LightsSchedule.propTypes = {
	schedule: PropTypes.node.isRequired,
	resetSchedule: PropTypes.func.isRequired,
	setSchedule: PropTypes.func.isRequired,
};

export {
	LightsSchedule,
};
