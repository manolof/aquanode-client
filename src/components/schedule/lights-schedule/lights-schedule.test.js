import { shallow } from 'enzyme';
import React from 'react';

import { LightsSchedule } from './lights-schedule';

describe('Lights Schedule Component', () => {
	const resetScheduleSpy = jest.fn();
	const setScheduleSpy = jest.fn();
	const mockSchedule = <p>Mock Schedule</p>;

	afterEach(() => {
		jest.resetAllMocks();
	});

	afterAll(() => {
		jest.restoreAllMocks();
	});

	it('should render', () => {
		const wrapper = shallow(
			<LightsSchedule schedule={mockSchedule} resetSchedule={resetScheduleSpy} setSchedule={setScheduleSpy}/>,
		);

		expect(wrapper.getElements()).toMatchSnapshot();
	});

	it(`should have functioning buttons that modify the lights' schedule`, () => {
		const wrapper = shallow(
			<LightsSchedule schedule={mockSchedule} resetSchedule={resetScheduleSpy} setSchedule={setScheduleSpy}/>,
		);

		expect(resetScheduleSpy).not.toHaveBeenCalled();
		wrapper.find('.schedule__reset').simulate('click');
		expect(resetScheduleSpy).toHaveBeenCalledTimes(1);

		wrapper.find('input[name="red"]').simulate('change', { target: { value: '199', name: 'red' }, persist: jest.fn() });
		expect(wrapper.state().red).toBe('199');

		expect(setScheduleSpy).not.toHaveBeenCalled();
		wrapper.find('.schedule__set').simulate('click');
		expect(setScheduleSpy).toHaveBeenCalledTimes(1);
		expect(setScheduleSpy).toHaveBeenLastCalledWith('lights', { blue: '', green: '', red: '199' });
	});
});
