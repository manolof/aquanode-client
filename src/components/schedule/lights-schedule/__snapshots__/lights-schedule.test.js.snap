// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lights Schedule Component should render 1`] = `
Array [
  <section
    className="schedule schedule__lights"
  >
    <h4>
      Lights
    </h4>
    <article
      className="schedule__values"
    >
      <p>
        Mock Schedule
      </p>
    </article>
    <form
      className="schedule__form"
    >
      <label>
        R
        <input
          max="255"
          min="0"
          name="red"
          onChange={[Function]}
          type="number"
          value=""
        />
      </label>
      <label>
        G
        <input
          max="255"
          min="0"
          name="green"
          onChange={[Function]}
          type="number"
          value=""
        />
      </label>
      <label>
        B
        <input
          max="255"
          min="0"
          name="blue"
          onChange={[Function]}
          type="number"
          value=""
        />
      </label>
      <label>
        W
        <input
          max="255"
          min="0"
          name="white"
          onChange={[Function]}
          type="number"
          value=""
        />
      </label>
    </form>
    <footer
      className="schedule__footer"
    >
      <button
        className="schedule__reset"
        onClick={[Function]}
      >
        RESET
      </button>
      <button
        className="schedule__set"
        onClick={[Function]}
      >
        SET
      </button>
    </footer>
  </section>,
]
`;
